snuba/datasets/entities/sessions.py

```diff
@@ -79,6 +91,127 @@ def plus_columns(
+
+
+has_errors = FunctionCall(
+    None, "greater", (Column(None, None, "errors"), Literal(None, 0))
+)
+did_not_nil = neq(distinct_id, lit_nil)
+duration_condition = _and(
+    neq(duration, Literal(None, MAX_UINT32)), eq(status, Literal(None, 1))
+)
+
+# These here are basically the same statements as the matview query
+sessions_raw_translators = TranslationMappers(
+    columns=[
+        ColumnToCurriedFunction(
+            None,
+            "duration_quantiles",
+            FunctionCall(None, "quantilesIf", quantiles),
+            (duration, duration_condition),
+        ),
+        ColumnToFunction(None, "duration_avg", "avgIf", (duration, duration_condition)),
+        ColumnToFunction(
+            None, "sessions", "sumIf", (quantity, eq(seq, Literal(None, 0))),
+        ),
+        ColumnToFunction(
+            None, "sessions_crashed", "sumIf", (quantity, eq(status, Literal(None, 2)))
+        ),
+        ColumnToFunction(
+            None, "sessions_abnormal", "sumIf", (quantity, eq(status, Literal(None, 3)))
+        ),
+        ColumnToFunction(
+            None,
+            "sessions_errored",
+            "plus",
+            (
+                FunctionCall(
+                    None,
+                    "uniqIf",
+                    (session_id, _and(has_errors, neq(session_id, lit_nil))),
+                ),
+                FunctionCall(
+                    None,
+                    "sumIf",
+                    (
+                        quantity,
+                        _and(
+                            _in(
+                                status, (Literal(None, status) for status in [2, 3, 4]),
+                            ),
+                            eq(session_id, lit_nil),
+                        ),
+                    ),
+                ),
+            ),
+        ),
+        ColumnToFunction(None, "users", "uniqIf", (distinct_id, did_not_nil)),
+        ColumnToFunction(
+            None,
+            "users_crashed",
+            "uniqIf",
+            (distinct_id, _and(eq(status, Literal(None, 2)), did_not_nil)),
+        ),
+        ColumnToFunction(
+            None,
+            "users_abnormal",
+            "uniqIf",
+            (distinct_id, _and(eq(status, Literal(None, 3)), did_not_nil)),
+        ),
+        ColumnToFunction(
+            None,
+            "users_errored",
+            "uniqIf",
+            (distinct_id, _and(has_errors, did_not_nil)),
+        ),
+    ]
+)
+
+
+class SessionsQueryStorageSelector(QueryStorageSelector):
+    def __init__(self) -> None:
+        self.materialized_storage = get_storage(StorageKey.SESSIONS_HOURLY)
+        self.raw_storage = get_storage(StorageKey.SESSIONS_RAW)
+
+    def select_storage(
+        self, query: Query, request_settings: RequestSettings
+    ) -> StorageAndMappers:
+        granularity = query.get_granularity() or 3600
+        use_materialized_storage = granularity >= 3600
```

Comment:
I don't think this is the right way to decide between the two storages.
By the time you get to this code `bucketed_started` will be already converted into into one of `toStartOfHour(starated)` or `toStartOfMinute(started)` or `toDate(started)` depending on the granularity field. At this point the granularity field in the query does not matter anymore. You need to look into what the query is grouping by.

Which makes me think something may be wrong in the sessions data model. If you query using `started` you are effectively querying by hour today without adding any bucketing. But then there is also a `bucketed_started` which applies the appropriate function according to the granularity. 
This is kind of wrong because the started column is already bucketed and we should not be adding another one. Which one do you use in your queries ?

Reply:
https://github.com/getsentry/sentry/blob/3c7e09479231b6e2c1fc4e78da2a2c270ede8d74/src/sentry/snuba/sessions_v2.py#L261
I use the bucketed_started.

Reply:
What would you recommend to do here? I don’t think matching on the generated AST is really feasible.

Reply:
Actually it seems you are using `bucketed_started` for the group by and the `started` for the time range of the query.
https://github.com/getsentry/sentry/blob/3c7e09479231b6e2c1fc4e78da2a2c270ede8d74/src/sentry/snuba/sessions_v2.py#L290-L291

I think the appropriate way to proceed is to use this conditions to decide whether to pick the aggregated table or the raw one:
Pick the aggregated table if:
- if the group by clause contains a hourly aggregation over the timestamp column (toStartOfHour(started)
- if the start and end are aligned to a start of hour
Otherwise pick the raw table and validates that the time range is not too long.

This implies also making the change on the product side to pass hourly aligned start and end as the time range. 

This way your data model would behave consistently providing the same result between raw and aggregated table. This should be a requirement for the consistency of the data model: if we decide to optimize the query by using the aggregated table, the result must be consistent with what you would get from the raw table. This means ensuring the start and end have to be aligned to the start of hour to decide to optimize.

In order to find this kind of information in the AST:
- This method gives you the time boundaries conditions of a query: https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/query_dsl/accessors.py#L93-L95. Takes a query and a column and returns to datetime.
- In order to find a specific expression in the AST we have a pattern matching api: https://github.com/getsentry/snuba/blob/master/snuba/query/matchers.py#L92. You can compose your pattern (that can have parameters) and that would match an expression tree. Then you can run it on all the expressions in the GROUP BY clause to find `toStartOfHour(started)`. There are a lot of examples in the test for matchers.py.

Did you run into other issues areound matching the AST ?

Reply:
> This implies also making the change on the product side to pass hourly aligned start and end as the time range.

That is already done. start/end are already rounded to a multiple of the `interval`.